### PR TITLE
gitlab: 16.11.5 -> 17.1.1

### PR DESCRIPTION
**Pull request by yayayayaka**
## Description of changes
- https://about.gitlab.com/releases/2024/05/16/gitlab-17-0-released/
- https://about.gitlab.com/releases/2024/06/20/gitlab-17-1-released/
- https://gitlab.com/gitlab-org/gitlab/-/blob/v17.1.1-ee/CHANGELOG.md

### BREAKING CHANGE: Disabling of runner registration tokens
Support for *runner registration tokens* is deprecated since GitLab 16.0, has been disabled by default in GitLab 17.0 and will be removed in GitLab 18.0, as outlined in the [GitLab documentation].

It is possible to [re-enable support for runner registration tokens] in the UI until GitLab 18.0, to prevent the registration workflow from breaking.

Support for *runner authentication tokens*, the replacement for registration tokens, has been implemented in #321052.
GitLab administrators should migrate to the [new runner registration workflow] with *runner authentication tokens* until the release of GitLab 18.0.

[GitLab documentation]: https://docs.gitlab.com/17.0/ee/ci/runners/new_creation_workflow.html#estimated-time-frame-for-planned-changes
[re-enable support for runner registration tokens]: https://docs.gitlab.com/17.0/ee/ci/runners/new_creation_workflow.html#prevent-your-runner-registration-workflow-from-breaking
[new runner registration workflow]: https://docs.gitlab.com/17.0/ee/ci/runners/new_creation_workflow.html#using-registration-tokens-after-gitlab-170

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking (#321052)
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by ajs124 — APPROVED** [review]

**Comment by ajs124** [comment]
Hm. This "breaks" gitlab-runners, registered through the nixos module.

https://docs.gitlab.com/17.0/ee/ci/runners/new_creation_workflow.html#using-registration-tokens-after-gitlab-170

So it seems like we'll finally need to get around to fixing https://github.com/NixOS/nixpkgs/issues/282238

File: pkgs/applications/version-management/gitlab/gitlab-shell/go-1_21.patch [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+index 513ccc4..0ecbdd2 100644
+--- a/go.mod
++++ b/go.mod
+@@ -1,6 +1,8 @@
+ module gitlab.com/gitlab-org/gitlab-shell/v14
+ 
+-go 1.20
++go 1.21
++
++toolchain go1.22.3
```
Comment (SuperSandro2000) [inline pkgs/applications/version-management/gitlab/gitlab-shell/go-1_21.patch:11]: Those lines are not strictly required and could cause the build to fail when go bitches about them.

**Comment by yayayayaka** [comment]
Thanks for pointing that out, ajs124. Unfortunately, I don't have the energy right now to push this through. Other maintainers are more than welcome to pick this up :purple_heart:

**Comment by nifoc** [comment]
Apart from changing the environment variable and possibly telling people to change [the setting](https://docs.gitlab.com/17.0/ee/administration/settings/continuous_integration.html#enable-runner-registrations-tokens) (until 18.0), what else is needed to get this in a mergeable state?

**Comment by dadada** [comment]
> Apart from changing the environment variable and possibly telling people to change [the setting](https://docs.gitlab.com/17.0/ee/administration/settings/continuous_integration.html#enable-runner-registrations-tokens) (until 18.0), what else is needed to get this in a mergeable state?

There are some options such as `--max-timeout` and `--tag-list` which the new `gitlab-runner register` (from 17.0 I think) will entirely refuse if `CI_SERVER_TOKEN` (this is the runner authorization token) is used instead of the legacy `REGISTRATION_TOKEN`. Afaik `--max-timeout` is always either 0 or some other value, so we would at least need to allow null and then not use the option, assuming the new runner registration workflow is used.

File: pkgs/applications/version-management/gitlab/gitlab-shell/go-1_21.patch (hunk 1 above)
Comment (yayayayaka) [reply pkgs/applications/version-management/gitlab/gitlab-shell/go-1_21.patch]: Good to know. Fortunately, the patch is no longer needed.

**Comment by yayayayaka** [comment]
Bumped to GitLab 17.1.0. `nixpkgs-review` happy, but the backup and restore test failed in `nixosTests.gitlab`.

**Comment by yayayayaka** [comment]
Switching back to `pkgs.nix` seems to have resolved my test issue. Putting it back into review.

@ofborg build nixosTests.gitlab

**Review by xanderio — APPROVED** [review]

**Review by SuperSandro2000 — APPROVED** [review]

**Pull request merged**